File: app/src/main/java/com/example/amazonbasicsirremote/AmazonBasicsCommands.java
package com.example.amazonbasicsirremote;

import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Map;

final class AmazonBasicsCommands {
    static final int MIN_TEMP_F = 62;
    static final int MAX_TEMP_F = 86;

    enum Mode { AUTO, COOL, DRY, FAN }
    enum FanSpeed { AUTO, LOW, MEDIUM, HIGH }

    private static final Map<String, int[]> COMMANDS = buildCommands();

    private AmazonBasicsCommands() {}

    static int[] forState(Mode mode, FanSpeed fanSpeed, int temperatureF) {
        String key = key(mode, fanSpeed, temperatureF);
        int[] command = COMMANDS.get(key);
        if (command == null) throw new IllegalArgumentException("Unsupported command: " + key);
        return Arrays.copyOf(command, command.length);
    }

    static int[] ledToggle() { return copy("LED_Toggle"); }
    static int[] energySaverToggle() { return copy("EnergySaver_Toggle"); }

    static Map<String, int[]> allCommands() { return COMMANDS; }

    private static int[] copy(String key) {
        int[] command = COMMANDS.get(key);
        return Arrays.copyOf(command, command.length);
    }

    private static String key(Mode mode, FanSpeed fanSpeed, int temperatureF) {
        if (mode == Mode.FAN) return "Fan_" + fanSpeed.name();
        if (temperatureF < MIN_TEMP_F || temperatureF > MAX_TEMP_F) {
            throw new IllegalArgumentException("Temperature must be 62-86°F");
        }
        return mode.name() + "_" + fanSpeed.name() + "_" + temperatureF + "F";
    }

    private static Map<String, int[]> buildCommands() {
        LinkedHashMap<String, int[]> map = new LinkedHashMap<>();
        addRange(map, Mode.AUTO, FanSpeed.AUTO, 130, new int[]{108,109,111,110,104,105,107,106,100,101,103,102,96,97,99,98,116,117,119,118,112,113,115,114,120});
        addRange(map, Mode.COOL, FanSpeed.AUTO, 160, new int[]{78,79,76,77,74,75,72,73,70,71,68,69,66,67,64,65,86,87,84,85,82,83,80,81,90});
        addRange(map, Mode.DRY, FanSpeed.AUTO, 129, new int[]{111,110,109,108,107,106,105,104,103,102,101,100,99,98,97,96,119,118,117,116,115,114,113,112,123});
        addRange(map, Mode.COOL, FanSpeed.LOW, 136, new int[]{102,103,100,101,98,99,96,97,106,107,104,105,108,109,111,110,122,123,120,121,124,125,127,126,114});
        addRange(map, Mode.COOL, FanSpeed.MEDIUM, 144, new int[]{118,119,116,117,114,115,112,113,122,123,120,121,124,125,127,126,102,103,100,101,98,99,96,97,106});
        addRange(map, Mode.COOL, FanSpeed.HIGH, 152, new int[]{122,123,120,121,124,125,127,126,114,115,112,113,116,117,119,118,106,107,104,105,108,109,111,110,98});
        map.put("Fan_AUTO", new int[]{161,164,126,255,255,91});
        map.put("Fan_LOW", new int[]{161,140,126,255,255,115});
        map.put("Fan_MEDIUM", new int[]{161,148,126,255,255,107});
        map.put("Fan_HIGH", new int[]{161,156,126,255,255,99});
        map.put("LED_Toggle", new int[]{162,8,255,255,255,117});
        map.put("EnergySaver_Toggle", new int[]{162,2,255,255,255,126});
        return Collections.unmodifiableMap(map);
    }

    private static void addRange(Map<String, int[]> map, Mode mode, FanSpeed fanSpeed, int modeFanByte, int[] checksums) {
        for (int i = 0; i < checksums.length; i++) {
            int temp = MIN_TEMP_F + i;
            map.put(key(mode, fanSpeed, temp), new int[]{161, modeFanByte, 96 + i, 255, 255, checksums[i]});
        }
    }
}

File: app/src/main/java/com/example/amazonbasicsirremote/AmazonBasicsCommands.kt
package com.example.amazonbasicsirremote

/**
 * Command-byte lookup table for the Amazon Basics/Midea air-conditioner IR captures.
 *
 * Temperature-bearing commands are available for 62F..86F. Fan-only and toggle
 * commands use the exact six-byte payloads supplied with the captures.
 */
object AmazonBasicsCommands {
    val TemperatureRange: IntRange = RemoteState.TemperatureRange

    val TemperatureCommands: Map<TemperatureCommandKey, IntArray> = buildMap {
        for (modeFan in TemperatureModeFan.entries) {
            for (temperatureF in TemperatureRange) {
                val key = TemperatureCommandKey(modeFan, temperatureF)
                put(key, RemoteState(modeFan.mode, modeFan.fanSpeed.toRemoteStateFanSpeed(), temperatureF).toCommandBytes())
            }
        }
    }

    val FanOnlyCommands: Map<FanOnlyCommandKey, IntArray> = mapOf(
        FanOnlyCommandKey(FanSpeed.AUTO) to intArrayOf(161, 164, 126, 255, 255, 91),
        FanOnlyCommandKey(FanSpeed.LOW) to intArrayOf(161, 140, 126, 255, 255, 115),
        FanOnlyCommandKey(FanSpeed.MEDIUM) to intArrayOf(161, 148, 126, 255, 255, 107),
        FanOnlyCommandKey(FanSpeed.HIGH) to intArrayOf(161, 156, 126, 255, 255, 99),
    )

    val ToggleCommands: Map<ToggleCommand, IntArray> = mapOf(
        ToggleCommand.LED_TOGGLE to intArrayOf(162, 8, 255, 255, 255, 117),
        ToggleCommand.ENERGY_SAVER_TOGGLE to intArrayOf(162, 2, 255, 255, 255, 126),
    )

    fun temperatureCommand(modeFan: TemperatureModeFan, temperatureF: Int): IntArray =
        TemperatureCommands.getValue(TemperatureCommandKey(modeFan, temperatureF)).copyOf()

    fun fanOnlyCommand(fanSpeed: FanSpeed): IntArray =
        FanOnlyCommands.getValue(FanOnlyCommandKey(fanSpeed)).copyOf()

    fun toggleCommand(toggle: ToggleCommand): IntArray =
        ToggleCommands.getValue(toggle).copyOf()
}

data class TemperatureCommandKey(
    val modeFan: TemperatureModeFan,
    val temperatureF: Int,
) {
    init {
        require(temperatureF in AmazonBasicsCommands.TemperatureRange) {
            "temperatureF must be in ${AmazonBasicsCommands.TemperatureRange}: $temperatureF"
        }
    }
}

data class FanOnlyCommandKey(val fanSpeed: FanSpeed)

enum class TemperatureModeFan(
    internal val mode: RemoteState.Mode,
    internal val fanSpeed: FanSpeed,
) {
    AUTO_AUTO(RemoteState.Mode.AUTO, FanSpeed.AUTO),
    COOL_AUTO(RemoteState.Mode.COOL, FanSpeed.AUTO),
    DRY_AUTO(RemoteState.Mode.DRY, FanSpeed.AUTO),
    COOL_LOW(RemoteState.Mode.COOL, FanSpeed.LOW),
    COOL_MEDIUM(RemoteState.Mode.COOL, FanSpeed.MEDIUM),
    COOL_HIGH(RemoteState.Mode.COOL, FanSpeed.HIGH),
}

enum class FanSpeed {
    AUTO,
    LOW,
    MEDIUM,
    HIGH,
}

enum class ToggleCommand {
    LED_TOGGLE,
    ENERGY_SAVER_TOGGLE,
}


private fun FanSpeed.toRemoteStateFanSpeed(): RemoteState.FanSpeed = when (this) {
    FanSpeed.AUTO -> RemoteState.FanSpeed.AUTO
    FanSpeed.LOW -> RemoteState.FanSpeed.LOW
    FanSpeed.MEDIUM -> RemoteState.FanSpeed.MEDIUM
    FanSpeed.HIGH -> RemoteState.FanSpeed.HIGH
}
